staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go

```diff
@@ -106,13 +142,33 @@ func HandleError(err error) {
 	}
 
 	for _, fn := range ErrorHandlers {
-		fn(err)
+		fn(context.Background(), err)
+	}
+}
+
+// HandlerError is a method to invoke when a non-user facing piece of code cannot
+// return an error and needs to indicate it has been ignored. Invoking this method
+// is preferable to logging the error - the default behavior is to log but the
+// errors may be sent to a remote server for analysis. The context is used to
+// determine how to log the error.
+//
+// This variant should be used instead of HandleError.
+func HandleErrorInContext(ctx context.Context, err error) {
+	// this is sometimes called with a nil error.  We probably shouldn't fail and should do nothing instead
+	if err == nil {
+		return
+	}
+
+	for _, fn := range ErrorHandlers {
+		fn(ctx, err)
 	}
 }
 
-// logError prints an error with the call stack of the location it was reported
-func logError(err error) {
-	klog.ErrorDepth(2, err)
+// logError prints an error with the call stack of the location it was reported.
+// It expects to be called as <caller> -> HandleError[InContext] -> logError.
+func logError(ctx context.Context, err error) {
+	logger := klog.FromContext(ctx).WithCallDepth(2)
+	logger.Error(err, "Unhandled error")
```

Comment:
I'm torn between calling this "Unhandled error" and "Handled error". I chose "Unhandled error" because logging an error is not the same as handling it in code - the whole point of logging at error level is that something unexpected occurred that needs attention.

Reply:
https://github.com/kubernetes/kubernetes/blob/bacac2547997d432fd6a93d7cef4aa64644544df/staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go#L411 has a test for this output.

Before the log entry was unstructured:
```
E1122 08:24:37.387820    9070 wrap.go:54] timeout or abort while handling: method=GET URI="/" audit-ID=""
```

Now it is structured:
```
E1121 21:17:48.088946   61801 wrap.go:54] "Unhandled error" err="timeout or abort while handling: method=GET URI=\"/\" audit-ID=\"\""
```

Did we intentionally or accidentally make this output part of the apiserver API? In other words, can this output be changed?

Would it make sense to change this further? There are [plenty of cases](https://grep.app/search?q=runtime.HandleError.%2Afmt.Errorf&regexp=true) where `HandleError` is called for an error that gets created via `fmt.Errorf` - this is unstructured printf-style logging! A `HandleErrorInContext(ctx, err, <key/value pairs>)` would lead to better output.

/cc @tkashem

Reply:
I like `UnhandledError` as this is what these usually are. The code variant of 🤷‍♂️, i.e. we cannot sensibly handle that error in the context, but want to log it for visibility that something is wrong.

Reply:
I've pushed an update. When passing message and key/value pairs as separate parameters, the "Unhandled error" has to be added via `WithName` because the message string is controlled by the called. The output in the timeout test now is:

```
"Timeout or abort while handling" logger="UnhandledError" method="GET" URI="/" auditID=""
```